prov/shm: set msg_order, mr mode, and complete with app context

aingerson opened the pull request:
- set cmd->msg.hdr.msg_id as app's inputted context for large messages so the tx can complete with that context on progress_resp, rather than just completing with a NULL context
- update MR mode to be backwards compatible with earlier versions of libfabric by adding FI_MR_SCALABLE
- add ability to switch to using virtual addresses based on app MR mode and rx msg_order restrictions, enabling "fast" RMA operations
- change tx_attr.comp_order to FI_ORDER_NONE

Signed-off-by: aingerson <alexia.ingerson@intel.com>

shefty commented on prov/shm/src/smr_attr.c:
```diff
@@ -68,7 +70,7 @@ struct fi_domain_attr smr_domain_attr = {
 	.resource_mgmt = FI_RM_ENABLED,
 	.av_type = FI_AV_UNSPEC,
-	.mr_mode = FI_MR_VIRT_ADDR,
+	.mr_mode = OFI_MR_BASIC_MAP | FI_MR_BASIC,
```
This doesn't look right.  Shm doesn't require memory registration for local buffers -- only for RMA target.  The app doesn't need to register send or receive buffers, or buffers that are used on the initiator side of RMA/atomic operations.

hppritcha commented:
ignore LANL CI check for now.

hppritcha commented:
bot:lanl:retest

aingerson commented:
Updated
- took out original mr_mode changes
- added offset support for backwards compatibility
- added enabling of FI_MR_VIRT_ADDR usage if app supports it, allowing for "fast" RMAs

shefty commented on prov/shm/src/smr_attr.c:
```diff
@@ -39,6 +39,7 @@
 struct fi_tx_attr smr_tx_attr = {
 	.caps = SMR_TX_CAPS,
 	.comp_order = FI_ORDER_STRICT,
+	.msg_order = FI_ORDER_STRICT,
```
FI_ORDER_STRICT is a completion ordering semantic, not msg ordering semantic. msg_order should be stuff like SAS, SAW, WAR, etc. send/write/read after send/write/read.  msg_order will be restricted by the use of the fast RMA protocol.  IIUC, fast RMA results in reads or writes being seen out of order at the receiver.

aingerson replied:
Actually, as it is right now I think fast RMA results would still be seen in order because the peer region is locked during the fast RMA so no one else can post to the receiver side until the CMA is over. Not optimal though. One of the areas I'm looking into is how to avoid this region locking during the memcpy but I think we can still keep it in order since the sender still needs a spot in the command queue and claims that before doing the copy.

shefty replied:
The problem is that fast RMA can be mixed with the non-fast RMAs.  The results is that a fast RMA can pass a 'slow' one which is sitting in the command queue.

aingerson replied:
Ahh yes. Got it. So it won't be able to ensure any RMA ordering. Is this going to be an issue with compatibility?

shefty replied:
Well, since you dynamically enable fast RMA based on support for MR_VIRT_ADDR, you could add msg_order to that selection as well.  If the app requests S/R/W after R/W or R/W after S/R/W (basically anything except SAS ordering -- I think), you can support it by disabling fast RMA.

shefty commented on prov/shm/src/smr_attr.c:
```diff
@@ -48,6 +49,7 @@ struct fi_tx_attr smr_tx_attr = {
 struct fi_rx_attr smr_rx_attr = {
 	.caps = SMR_RX_CAPS,
 	.comp_order = FI_ORDER_STRICT,
+	.msg_order = FI_ORDER_SAS | FI_ORDER_STRICT,
```
See comment for tx_attr.

shefty commented on prov/shm/src/smr_ep.c:
```diff
@@ -251,20 +251,20 @@ void smr_format_inject(struct smr_cmd *cmd, fi_addr_t peer_id,
 	cmd->msg.data.iov_count = count;
 	cmd->msg.hdr.size = total_len;
-	cmd->msg.hdr.msg_id = (uint64_t) (uintptr_t) pend_cmd;
+	cmd->msg.hdr.msg_id = (uint64_t) context;
```
This needs the (uintptr_t) cast.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -54,7 +54,7 @@ static void smr_progress_resp(struct smr_ep *ep)
 		pending = (struct smr_cmd *) resp->msg_id;
 
-		ret = ep->tx_comp(ep, NULL,
+		ret = ep->tx_comp(ep, (void *) pending->msg.hdr.msg_id,
```
This needs the (uintptr_t) cast as well.

shefty commented on prov/shm/src/smr_init.c:
```diff
@@ -68,6 +68,9 @@ static int smr_getinfo(uint32_t version, const char *node, const char *service,
 	int ret;
 
+	if (hints && hints->domain_attr->mr_mode & FI_MR_VIRT_ADDR)
+		smr_util_prov.info->domain_attr->mr_mode |= FI_MR_VIRT_ADDR;
```
Note that this isn't thread safe.

shefty commented on prov/shm/src/smr_rma.c:
```diff
@@ -145,9 +121,13 @@ ssize_t smr_generic_rma(struct fid_ep *ep_fid, const struct iovec *iov,
 
+	if (ep->util_ep.domain->mr_mode & FI_MR_VIRT_ADDR &&
+	    rma_count == 1 && !(op_flags & OFI_REMOTE_CQ_DATA))
+		cmds = 1;
```
You can use:
```
cmds = 1 +( (mr_mode & VIRT_ADDR) && (rma_count == 1) && !(op_flags & CQ_DATA)
```
This eliminates the need to initialize cmds and the branch from the if statement.  Same comment applies further down

shefty commented:
It would be best to keep this as a single provider.  The provider structure allows for a linked-list of fi_info structures, so you could have separate fi_info's for the fast and strict cases without needing to plug in as different providers.

Based on the attributes specified when opening the endpoint, you could select a different set of RMA calls optimized for either the fast or strict case.

aingerson commented:
Ok, updated again. I decided just to manually alter the MR mode and order if needed and then added a "fast_rma" field per domain to reflect enabling of the fast path. Setting different functions by endpoint would be a lot of code duplication and a hassle. I think this is simpler.

shefty commented on prov/shm/src/smr_init.c:
```diff
@@ -66,8 +66,12 @@ static int smr_getinfo(uint32_t version, const char *node, const char *service,
 	int ret;
 
+	fast_rma = hints ? smr_fast_rma_enabled(hints->domain_attr->mr_mode,
+						hints->tx_attr->msg_order) : 0;
```
merging, but please look at this test.  There's the potential to de-reference a null pointer.  Also, I think you can enable fast_rma if no hints are provided.  I think the correct check sequence would be:
```
mr_mode = hints && hints->domain_attr ? hints->domain-attr : FI_MR_VIRT_ADDR;
msg_order = hints && hints->tx_attr ? hints->tx_attr->msg_order : FI_ORDER_SAS;

fast_rma = smr_fast_rma_enabled(mr_mode, msg_order);
```
see below
